pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -0,0 +1,539 @@
+		klog.V(2).InfoS("Error syncing Service, retrying", "service", key, "err", err)
+		r.ipQueue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Dropping Service %q out of the queue: %v", key, err)
+	r.ipQueue.Forget(key)
+	runtime.HandleError(err)
+}
+
+// syncIPAddress verify that the IPAddress that are owned by the ipallocator controller reference an existing Service
+// to avoid leaking IPAddresses. IPAddresses that are owned by other controllers are not processed to avoid hotloops.
+// IPAddress that reference Services and are part of the ClusterIP are validated in the syncService loop.
+func (r *RepairIPAddress) syncIPAddress(key string) error {
+	ipAddress, err := r.ipAddressLister.Get(key)
+	if err != nil {
+		// nothing to do
+		return nil
+	}
+
+	// not mananged by this controller
+	if !managedByController(ipAddress) {
+		return nil
+	}
+
+	// does not reference a Service but created by the service allocator, something else have changed it, delete it
+	if ipAddress.Spec.ParentRef.Group != "" ||
+		ipAddress.Spec.ParentRef.Resource != "services" {
+		runtime.HandleError(fmt.Errorf("IPAddress %s appears to have been modified, not referencing a Service %v: cleaning up", ipAddress.Name, ipAddress.Spec.ParentRef))
+		r.recorder.Eventf(ipAddress, nil, v1.EventTypeWarning, "IPAddressNotAllocated", "IPAddressAllocation", "IPAddress %s appears to have been modified, not referencing a Service %v: cleaning up", ipAddress.Name, ipAddress.Spec.ParentRef)
+		err := r.client.NetworkingV1alpha1().IPAddresses().Delete(context.Background(), ipAddress.Name, metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	svc, err := r.serviceLister.Services(ipAddress.Spec.ParentRef.Namespace).Get(ipAddress.Spec.ParentRef.Name)
+	if apierrors.IsNotFound(err) {
+		// cleaning all IPAddress without an owner reference IF the time since it was created is greater than 60 seconds (default timeout value on the kube-apiserver)
+		// This is required because during the Service creation there is a time that the IPAddress object exists but the Service is still being created
+		// Assume that CreationTimestamp exist
+		ipLifetime := r.clock.Now().Sub(ipAddress.CreationTimestamp.Time)
+		gracePeriod := 60 * time.Second
+		if ipLifetime > gracePeriod {
+			runtime.HandleError(fmt.Errorf("IPAddress %s appears to have leaked: cleaning up", ipAddress.Name))
+			r.recorder.Eventf(ipAddress, nil, v1.EventTypeWarning, "IPAddressNotAllocated", "IPAddressAllocation", "IPAddress: %s for Service %s/%s appears to have leaked: cleaning up", ipAddress.Name, ipAddress.Spec.ParentRef.Namespace, ipAddress.Spec.ParentRef.Name)
+			err := r.client.NetworkingV1alpha1().IPAddresses().Delete(context.Background(), ipAddress.Name, metav1.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+		// requeue after the grace period
+		r.ipQueue.AddAfter(key, gracePeriod-ipLifetime)
```

Comment:
`processNextWorkIp()` operates on each IP, calling  `handleIpErr` which re-Q in case of error. Why are we re-Qing here?

Reply:
when you create a Service PUT /service
1. internally you create an IP PUT /ipaddress
2. something happen here and the apiserver takes a bit of time to create the service
3. if the repair loop sees the IPAddress before the Service is created it will think that is orphan

to avoid any problem, I just queue to check it after the grace period, if the apiserver was slow the Service should exist, if the Service failed to be created or just crashed, this IP will be orphan